designs/sam-config.md

```diff
@@ -0,0 +1,282 @@
+
+If this could be condensed into a series of workflows that look like
+
+* `sam build`
+* `sam package`
+* `sam deploy`
+
+That would be a huge user experience win.
+
+What will be changed?
+---------------------
+
+The suite of commands supported by SAM CLI would be aided by looking for a configuration file thats locally located at the project root where template.yaml is located by default. 
+
+`.sam-app-config`
+
+This configuration would be used for specifiying the parameters that each of SAM CLI commands use and would be in TOML format.
+
+Running a SAM CLI command now automatically looks for `.sam-config` file and if its finds it goes ahead with parameter passthroughs to the CLI.
+
+```
+sam build
+Default Config file location: .sam-config
+..
+..
+..
+```
+
+
+The default location of a .sam-config can be replaced by overriding an environment variable called `SAM_CONFIG`
+
+`
+export SAM_CONFIG=~/Users/username/mysamconfig
+`
+
+Users can pass an identifier for the section that will be scanned within the configuration file to pass parameters through.
+
+Be default the `default` section of the configuration is chosen.
+
+```
+[default]
+
+[default.build]
+[default.build.paramaters]
+profile="srirammv"
+debug=true
+skip_pull_image=true
+use_container=true
+
+[default.package]
+[default.package.parameters]
+profile="srirammv"
+region="us-east-1"
+s3_bucket="sam-bucket"
+output_template_file="packaged.yaml"
+
+[default.deploy]
+[default.deploy.parameters]
+stack_name="using_config_file"
+capabilities="CAPABILITY_IAM"
+region="us-east-1"
+profile="srirammv"
+
+```
+
+If a custom identifer is specified, the identifier is looked up `.sam-config` file instead.
+
+`sam build --identifier dev`
+
+Sample configuration file
+
+```
+[default]
+
+[default.build]
+[default.build.paramaters]
+profile="srirammv"
+debug=true
+skip_pull_image=true
+use_container=true
+
+[default.package]
+[default.package.parameters]
+profile="srirammv"
+region="us-east-1"
+s3_bucket="sam-bucket"
+output_template_file="packaged.yaml"
+
+[default.deploy]
+[default.deploy.parameters]
+stack_name="using_config_file"
+capabilities="CAPABILITY_IAM"
+region="us-east-1"
+profile="srirammv"
+
+[dev]
+
+[dev.build]
+[dev.build.paramaters]
+profile="srirammv"
+debug=true
+skip_pull_image=true
+use_container=true
+
+[dev.package]
+[dev.package.parameters]
+profile="srirammv"
+region="us-east-1"
+s3_bucket="sam-bucket"
+output_template_file="packaged.yaml"
+
+[dev.deploy]
+[dev.deploy.parameters]
+stack_name="using_config_file"
+capabilities="CAPABILITY_IAM"
+region="us-east-1"
+profile="srirammv"
+```
+
+
+The configuration file can then be potentially intialized
+
+* all sam init projects could come with a sample .sam-config file
+
+Open Questions
+-------------------------------
+
+* Potentially every sam command could have functionality to have a series of command line parameters exported into a configuraion file.
+
+Tenets
```

Comment:
Nit: consider moving this up before "What will be changed?". Might help set some context and be easier to reference and update when we make changes in the future.